File: rust/src/protocol/v763/packets/play/clientbound/scoreboard_score.rs
//! Packet `scoreboard_score` (play/clientbound). Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot, string_codec as string, varint,
    varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ClientboundPacket;

pub const PACKET_ID: i32 = 0x5B;

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ScoreboardScore {
    pub item_name: String,
    pub action: i32,
    pub score_name: String,
    pub value: Option<i32>,
}

impl ScoreboardScore {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let item_name = string::read(reader)?;
        let action = varint::read(reader)?;
        let score_name = string::read(reader)?;
        let value = if action == 1 {
            None
        } else {
            Some(varint::read(reader)?)
        };
        Ok(Self {
            item_name,
            action,
            score_name,
            value,
        })
    }
}

impl ClientboundPacket for ScoreboardScore {
    fn state(&self) -> ConnectionState {
        ConnectionState::Play
    }
    fn packet_id(&self) -> i32 {
        PACKET_ID
    }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        string::write(&self.item_name, writer)?;
        varint::write(self.action, writer)?;
        string::write(&self.score_name, writer)?;
        if let Some(v) = self.value {
            varint::write(v, writer)?;
        }
        Ok(())
    }
}

File: python/minecraft_bot/protocol/v763/packets/play/clientbound/scoreboard_score.py
"""Packet `scoreboard_score` (play/clientbound, id 0x5B).

Updates a scoreboard score. ``action == 1`` (remove) omits ``value``;
otherwise ``value`` is a varint.
"""

from __future__ import annotations

from dataclasses import dataclass

from minecraft_bot.codec import Reader, Writer, string, varint

PACKET_ID = 0x5B


@dataclass(frozen=True, slots=True)
class ScoreboardScore:
    item_name: str          # entity / player name
    action: int             # varint: 0=update/create, 1=remove
    score_name: str         # objective name
    value: int | None    # varint, present when action != 1


def decode(reader: Reader) -> ScoreboardScore:
    item = string.read(reader)
    act = varint.read(reader)
    sn = string.read(reader)
    if act == 1:
        val: int | None = None
    else:
        val = varint.read(reader)
    return ScoreboardScore(item_name=item, action=act, score_name=sn, value=val)


def encode(packet: ScoreboardScore, writer: Writer) -> None:
    string.write(packet.item_name, writer)
    varint.write(packet.action, writer)
    string.write(packet.score_name, writer)
    if packet.action != 1:
        if packet.value is None:
            from minecraft_bot.errors import ValueOutOfRange
            raise ValueOutOfRange("scoreboard_score.value", packet.value)
        varint.write(packet.value, writer)
